qkernel/src/lib.rs

```diff
@@ -483,20 +517,47 @@ cfg_if::cfg_if! {
             vdsoParamAddr: u64,
             vcpuCnt: u64,
             autoStart: bool,
+            vm_type: u64
         ) {
 
             self::qlib::kernel::asm::fninit();
             if id == 0 {
-
+                let vm_type = VmType::from_u64(vm_type).unwrap_or_default();
+                match vm_type {
+                    VmType::SevSnp =>{
+                        ENABLE_CC.store(true, Ordering::Release);
+                        IS_SEV_SNP.store(true, Ordering::Release);
+                        //Log is available after ghcb is initialized
+                        LOG_AVAILABLE.store(false, Ordering::Release);
+                        for i in (MemoryDef::phy_lower_gpa()..MemoryDef::guest_host_shared_heap_end_gpa()).step_by(MemoryDef::PAGE_SIZE as usize) {
+                            let _ret = pvalidate(VirtAddr::new(i), PvalidateSize::Size4K, true);
+                        }
+                    },
+                    VmType::CCEmu =>{
+                        ENABLE_CC.store(true, Ordering::Release);
+                    }
+                    _ =>(),
+                }
 
                 GLOBAL_ALLOCATOR.InitPrivateAllocator(MemoryDef::guest_private_running_heap_offset_gpa());
+                unsafe {KERNEL_PAGETABLE.Init(PageTables::Init(CurrentKernelTable()&0xffff_ffff_ffff));}
+
+                //set idt first here cpuid is interceptted in cc
+                unsafe{
+                    interrupt::InitSingleton();
+                }
+                interrupt::init();
+
+                match vm_type {
+                    VmType::SevSnp =>{
+                        set_cbit_mask();
+                        PAGE_MGR.SetValue(PAGE_MGR_HOLDER.Addr());
+                        // ghcb convert shared memory
+                        InitShareMemory();
+                    },
+                    _ =>(),
+                }
 
-                
-    
-                //check msr to see if sev-snp enabled
-                ENABLE_CC.store(true,Ordering::Release);
-                // ghcb convert shared memory
-                
                 GLOBAL_ALLOCATOR.InitSharedAllocator(MemoryDef::guest_host_shared_heap_offest_gpa());
```

Comment:
not related to this PR, but I'm still not convinced of having getters for static constant values. This only makes code reading harder for me.